File: daily-coding-problem/dcp-jvm/dcp-scala/src/main/scala/dcp/problems/_1_20/Problem1_PairsWithSum.scala
package dcp.problems._1_20

/**
 * This problem was asked by Google.
 *
 * Given a list of numbers and a number k, return whether any two numbers from the list add up to k.
 *
 * For example, given [10, 15, 3, 7] and k of 17, return true since 10 + 7 is 17.
 */

object Problem1_PairsWithSum {

  // brute force O(n^2)
  def solution(arr: Array[Int], sum: Int): Boolean = {
    for (i <- arr.indices) {
      val acc = sum - arr(i)
      for (j <- (i + 1) until arr.length) {
        if (acc - arr(j) == 0) return true
      }
    }
    false
  }

  // using complements O(n) - single pass
  def solution2(arr: Array[Int], sum: Int): Boolean = {
    var complements = Set.empty[Int]
    for (i <- arr.indices) {
      if (complements.contains(arr(i))) return true
      complements += sum - arr(i)
    }
    false
  }


}

File: daily-coding-problem/dcp-py/1_20/problem1_pairswithsum.py
"""
This problem was recently asked by Google.

Given a list of numbers and a number k, return whether any two numbers from the list add up to k.

For example, given [10, 15, 3, 7] and k of 17, return true since 10 + 7 is 17.

Bonus: Can you do this in one pass?
"""


def pairswithsum(input, k):
    comps = set()
    for e in input:
        if e in comps:
            return True
        comps.add(k - e)
    return False


assert (pairswithsum([10, 15, 3, 7], 17))  # 10 + 7
assert (pairswithsum([12, 4, 6, 9, 15, 13, 7], 25))  # 12 + 13
